README.md
# MashuJS

MashuJS is a command handler for [discord.js](https://discord.js.org/#/) written in TypeScript for [astolto](https://github.com/EpokTarren/astolto).

# Getting started

The client provided is an extension of discord.js [client](https://discord.js.org/#/docs/main/stable/class/Client) that has a handler attached. You can also view the [documentation page](https://mashu.tarren.moe/) for more extensive docs.

```sh
# with npm
npm install mashujs discord.js
# with yarn
yarn add mashujs discord.js
```

Initialise a client and pass it to a handler

```js
const { Client } = require('mashujs');
const { resolve } = require('path');

const client = new Client({
	disableMentions: 'everyone',
	dir: resolve(__dirname, 'commands'),
	owners: ['your id'],
	prefix: 'a;',
	enableHelp: true,
	errorChannel: 'Some channel id',
});

client.once('ready', () => {
	client.handler.loadSlashCommands();
});

client.login('your bot token');
```

Simple example command

```ts
const max = 1 << 32;
export = new Command({
	async run(message) {
		let n = message.options.getInteger('n') || 6;

		if (n > max) n = max;
		else if (n < 2) n = 6;

		const roll = Math.floor(Math.random() * n) + 1;

		message.reply(`You rolled ${roll}`);
	},
	name: 'Roll',
	aliases: ['R', 'Dice', 'D'],
	description: 'Roll an n-sided die.',
	detailed: 'Roll a number between 1 and n.',
	examples: [(prefix) => `${prefix}roll`, (prefix) => `${prefix}r 1000`, (prefix) => `${prefix}d 1000000`],
	arguments: [
		{
			type: 'Integer',
			name: 'n',
			description: 'Max roll value.',
			required: false,
		},
	],
	parseArgs: true,
	interaction: 'on',
});
```

You can set some environment variables to change the styling of output

```ts
process.env.HELPFOOTER; // Footer text for help command
process.env.HELPFOOTERICON; // Footer icon for help command
process.env.MASHUCOLOR; // Default color for built in commands default "0xff80cc"
process.env.MASHUERRORCOLOR; // Error color for built in error logging default "0xff8080"
process.env.MASHUDEBUGGUILD; // Guild to limit testing of slash command pushes to, default is undefined
```

# Docs generation

Default `.mashurc.json` config can be generated by running

```sh
npx mashujs --new
```

After that edit your config.

## Properties

- `prefix`, Prefix to use when generating examples.
- `input`, Path to command locations.
- `output`, Path to docs folder.
- `readme`, Path to the readme.md for the bot as a whole and will export as index.md.
- `descriptions`, Will include a description beneath the title if one exists in the command categories folder as README.md.
- `descriptionReplacer`, Replace build path with src path to locate README files, ["build", "src"]
- `titles`, Whether or not to include jekyll titles.
- `permalink`, Whether or not to include jekyll permalinks.
- `layout`, Jekyll layout for command pages.
- `homeHeader`, Jekyll header options for index.md such as; "layout: home".

# License

MashuJS is licensed under [MIT](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 2.0.1

Slash command improvements.

### Added

- When testing commands, you may set `process.env.MASHUDEBUGGUILD` so that commands are only registered in your test guild.

### Changed

- Registering slash commands no longer needs a token passed in.
- @discordjs/builders, @discordjs/rest, and discord-api-types are no longer required.

### Removed

- Superfluous console logs when each command loads.

## 2.0.0

discord.js v13 migration and slash command support.

### Added

- SlashCommand support
  - `MessageCommand` for message only commands.
  - `IntractableCommand` for message or slash command support in the same file.
  - `InteractionCommand` for slash command only commands.
- Messages now have an added args property which will always be a string array.
- Command argument parsing for message commands with the same API as discord.js interactions.

### Changed

- BaseCommand should be replaced with `MessageCommand`, `IntractableCommand`, or `InteractionCommand`, depending on if a command is message only, message and interaction or only interaction based.
- `IntractableCommand`, and `InteractionCommand` type functions don't get arguments passed in but `IntractableCommand` if the command invocation is a message args will be a property of the message.
- Migrated to discord.js v13 this means Node.js 16.6.0 or newer is required.
- Fixed bug where a non string description didn't throw an error.
- User now needs to install @discordjs/builders @discordjs/rest discord-api-types in addiction to discord.js.

## 1.0.1 - 2021-04-15

### Added

- TSDoc annotations

### Changed

- Documentation links no longer include leading `/classes/` or `/interfaces/`.
- Documentation now links to the license file in the repository.
- Fix non relative links being mangled.
- Link to documentation in readme.
- Updated the spelling of typescript to TypeScript.
- Command reloading now does not error because of duplicate names.
- Reload now responds with "Unable to reload command." if an owner reloads a command it cannot reload or that does not exist.
- Fix npx script not working.

## 1.0.0 - 2021-04-10

Initial release.


src/command.ts
import { Permissions, PermissionString } from 'discord.js';
import { Message, Interaction, IntractableMessage } from './client';

/**
 * Resolvable to an example, either a string or a function that resolves to a string.
 */
export type ExampleResolvable = string | ((prefix: string) => string);

export type ArgumentType = 'Boolean' | 'Channel' | 'Integer' | 'Mentionable' | 'Role' | 'String' | 'User';

export interface Argument {
	name: string;
	description: string;
	type: ArgumentType;
	required?: boolean;
	choices?: { name: string; value: string | number }[];
}

export interface BaseCommand {
	/**
	 * Commands execution function.
	 */
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	run: (msg: any, args: string[]) => Promise<void> | void;

	/**
	 * Name of the command.
	 */
	name: string;

	/**
	 * Short description of the command.
	 */
	description: string;

	/**
	 * Whether the command is hidden.
	 */
	hidden?: boolean;

	/**
	 * Aliases of the command.
	 */
	aliases?: string[];

	/**
	 * Category the command is in.
	 */
	category?: string;

	/**
	 * Detailed description of the command.
	 */
	detailed?: string;

	/**
	 * Examples of the command in use.
	 */
	examples?: ExampleResolvable[];

	/**
	 * If the command is available only in guilds.
	 */
	guildOnly?: boolean;

	/**
	 * Whether the command is suitable to be a message interaction.
	 */
	interaction: 'off' | 'on' | 'only';

	/**
	 * Command arguments
	 */
	arguments?: Argument[];

	/**
	 * Whether to parse arguments for plain message command invocations.
	 */
	parseArgs?: boolean;

	/**
	 * Permissions needed to use the command.
	 */
	permissions?: PermissionString[];

	/**
	 * Permissions the bot needs to execute the command.
	 */
	botPermissions?: PermissionString[];
}

export interface MessageCommand extends BaseCommand {
	run: (msg: Message, args: string[]) => Promise<void> | void;
	interaction: 'off';
}

/**
 * A command that is optionally a slash command
 */
export interface IntractableCommand extends BaseCommand {
	/**
	 * Commands execution function.
	 */
	run: (msg: IntractableMessage) => Promise<void> | void;

	interaction: 'on';
}

/**
 * A command that is exclusively a slash command
 */
export interface InteractionCommand extends BaseCommand {
	/**
	 * Commands execution function.
	 */
	run: (msg: Interaction) => Promise<void> | void;

	interaction: 'only';

	parseArgs?: never;
}

const commaReduce = (acc: string, perm: string) => `${acc}, ${perm}`;
const listReduce = (acc: string, value: string) => `${acc}• ${value}\n`;
const legible = (perm: PermissionString) =>
	perm
		.toLowerCase()
		.replace('_', ' ')
		.replace(/(^| )./g, (str) => str.toUpperCase());

/**
 * The metadata of a command.
 */
export interface CommandMetadata {
	/**
	 * Name of the command.
	 */
	readonly name: string;

	/**
	 * Short description of the command.
	 */
	readonly description: string;

	/**
	 * Aliases of the command.
	 */
	readonly aliases: string[];

	/**
	 * Category the command is in.
	 */
	readonly category: string;

	/**
	 * Detailed description of the command.
	 */
	readonly detailed: string;

	/**
	 * Examples of the command in use.
	 */
	readonly examples: string[];

	/**
	 * If the command is available only in guilds.
	 */
	readonly guildOnly: boolean;

	/**
	 * Whether the command is suitable to be a message interaction.
	 */
	readonly interaction: 'off' | 'on' | 'only';

	/**
	 * Interaction arguments
	 */
	readonly arguments: Argument[];

	/**
	 * Permissions needed to use the command as human readable.
	 */
	readonly permissions: string[];

	/**
	 * Permissions the bot needs as human readable.
	 */
	readonly botPermissions: string[];
}

/**
 * Represents a command.
 * @example
 * ```ts
 * const max = Math.pow(2, 32);
 * export = new Command({
 * 	async run(message) {
 * 		let n = message.options.getInteger('n') || 6;
 *
 * 		if (n > max) n = max;
 * 		else if (n < 2) n = 6;
 *
 * 		const roll = Math.floor(Math.random() * n) + 1;
 *
 * 		message.reply(`You rolled ${roll}`);
 * 	},
 * 	name: 'Roll',
 * 	aliases: ['R', 'Dice', 'D'],
 * 	description: 'Roll an n-sided die.',
 * 	detailed: 'Roll a number between 1 and n.',
 * 	examples: [
 * 		(prefix) => `${prefix}roll`,
 * 		(prefix) => `${prefix}r 1000`,
 * 		(prefix) => `${prefix}d 1000000`,
 * 	],
 * 	arguments: [
 * 		{
 * 			type: 'Integer',
 * 			name: 'n',
 * 			description: 'Max roll value.',
 * 			required: false,
 * 		},
 * 	],
 * 	parseArgs: true,
 *	interaction: 'on',
 * });
 *
 * ```
 */
export class Command {
	/**
	 * Commands execution function.
	 */
	public readonly run: (msg: IntractableMessage, args: string[]) => Promise<void>;

	/**
	 * Name of the command.
	 */
	public readonly name: string;

	/**
	 * Short description of the command.
	 */
	public readonly description: string;

	/**
	 * Help text describing the command.
	 */
	public readonly help: string;

	/**
	 * Short help text for summary viewing of the command.
	 */
	public readonly shortHelp: string;

	/**
	 * Aliases of the command.
	 */
	public readonly aliases: string[];

	/**
	 * Category the command is in.
	 */
	public readonly category: string;

	/**
	 * Detailed description of the command.
	 */
	public readonly detailed: string;

	/**
	 * Examples of the command in use.
	 */
	public readonly examples: string[];

	/**
	 * Whether the command is hidden.
	 */
	public readonly hidden: boolean;

	/**
	 * If the command is available only in guilds.
	 */
	public readonly guildOnly: boolean;

	/**
	 * Whether the command is suitable to be a message interaction.
	 */
	public readonly interaction: 'off' | 'on' | 'only';

	/**
	 * Interaction arguments
	 */
	public readonly arguments: Argument[];

	/**
	 * Whether to parse arguments for plain message command invocations.
	 */
	public readonly parseArgs: boolean;

	/**
	 * Permissions needed to use the command.
	 */
	public readonly permissions: Permissions;
	/**
	 * Permissions the bot needs to execute the command.
	 */
	public readonly botPermissions: Permissions;

	/**
	 * Permissions needed to use the command as a string.
	 */
	public readonly permissionsText: string;
	/**
	 * Permissions needed to execute the command as a string.
	 */
	public readonly botPermissionsText: string;

	/**
	 * Get the metadata of a command.
	 * @returns Metadata of the command.
	 */
	public metadata(): CommandMetadata {
		return {
			name: this.name,
			description: this.description,
			aliases: this.aliases,
			category: this.category,
			detailed: this.detailed,
			examples: this.examples,
			guildOnly: this.guildOnly,
			interaction: this.interaction,
			arguments: this.arguments,
			permissions: this.permissionsText.split(', '),
			botPermissions: this.botPermissionsText.split(', '),
		};
	}

	/**
	 * Creates an instance of a command.
	 * @param command - The command resolvable.
	 * @param prefix - Prefix that is currently in use.
	 * @param directory - Name of the parent folder.
	 */
	constructor(
		command: MessageCommand | IntractableCommand | InteractionCommand | Command,
		prefix = '${prefix}',
		directory = ''
	) {
		const {
			run,
			name,
			description,
			hidden = false,
			aliases = [],
			category = '',
			detailed = '',
			examples = [],
			guildOnly = false,
			interaction = 'off',
			parseArgs = false,
			permissions = [],
			botPermissions = [],
		} = command;

		if (typeof run !== 'function') throw new TypeError('run is not a function.');

		if (typeof name !== 'string') throw new TypeError('Command name must be a string.');

		if (typeof description !== 'string') throw new TypeError('Command description must be a string.');

		this.run =
			run.constructor.name === 'AsyncFunction'
				? (run as Command['run'])
				: async (m: IntractableMessage, a: string[]) => (run as Command['run'])(m, a);

		this.name = name;

		this.description = description;

		this.aliases = aliases;

		this.category = directory || category;

		this.detailed = detailed || description;

		this.examples = examples.map((example) =>
			typeof example === 'string' ? example.replace('${prefix}', prefix) : example(prefix)
		);

		this.hidden = hidden;

		this.guildOnly = guildOnly;

		this.interaction = interaction;

		this.arguments = command.arguments || [];

		this.parseArgs = parseArgs;

		if (command.constructor.name !== 'Command') {
			this.permissions = new Permissions(permissions);

			this.botPermissions = new Permissions(botPermissions);

			this.permissionsText = this.permissions.bitfield
				? (permissions as PermissionString[]).map(legible).reduce(commaReduce).trim()
				: '';
			this.botPermissionsText = this.botPermissions.bitfield
				? (botPermissions as PermissionString[]).map(legible).reduce(commaReduce).trim()
				: '';
		} else {
			this.permissionsText = (command as Command).permissionsText;
			this.botPermissionsText = (command as Command).botPermissionsText;
			this.permissions = (command as Command).permissions;
			this.botPermissions = (command as Command).botPermissions;
		}

		this.help =
			`${description}\n` +
			(this.aliases[0] ? `**Aliases**\n${this.aliases.reduce(listReduce, '')}` : '') +
			(this.examples
				? `**Examples**\n${this.examples
						.map((v) => `\`${v}\``)
						.reduce(listReduce, '')
						.trim()}`
				: '') +
			(this.permissions.bitfield ? `\n**Permissions**${this.permissions.toArray().reduce(listReduce, '')}` : '') +
			(this.botPermissions.bitfield
				? `\n**Bot Permissions**${this.botPermissions.toArray().reduce(listReduce, '')}`
				: '');

		this.shortHelp = `**${this.name}**: ${description}`;
	}
}
